### PR TITLE
HPCC-24392 Loading Roxie Query graph on ECLWatch locks roxie queries

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@mckellyln @ghalliday Please review

**Comment by richardkchapman** [comment]
Retargeted to 7.6.x

File: roxie/ccd/ccdstate.cpp [hunk 1]
```diff
@@ -1280,13 +1280,68 @@ static hash64_t hashXML(const IPropertyTree *tree)
...
+    {
+        CriticalBlock b2(updateCrit);
+        return slaveManagers.getLink();
+    }
+
+protected:
+
+    // Derived classes wanting to read serverManager or slaveManagers must call this function to safely obtain their current values
+
+    void getQueryManagers(Owned<IRoxieQuerySetManager> _serverManager, Owned<CRoxieSlaveQuerySetManagerSet> &_slaveManagers) const
```
Comment (ghalliday) [inline roxie/ccd/ccdstate.cpp:1308]: Should be a reference parameter... otherwise it won't get modified...

File: roxie/ccd/ccdstate.cpp [hunk 2]
```diff
@@ -1280,13 +1280,68 @@ static hash64_t hashXML(const IPropertyTree *tree)
...
+        }
+        if (slaveQueryReleaseDelaySeconds)
+            delayedReleaser->delayedRelease(oldSlaveManagers.getClear(), slaveQueryReleaseDelaySeconds);
+    }
+
+private:
+    mutable CriticalSection updateCrit;  // protects updates of slaveManagers and serverManager, and queryHash. Must be held ONLY to link, release, or overwrite these values.
+    Owned<CRoxieSlaveQuerySetManagerSet> slaveManagers;
+    Owned<IRoxieQuerySetManager> serverManager;
+    hash64_t queryHash;
```
Comment (ghalliday) [inline roxie/ccd/ccdstate.cpp:1336]: no longer initialised to 0

**Comment by ghalliday** [comment]
@richardkchapman a couple of problems.

**Comment by richardkchapman** [comment]
@ghalliday Fixed and repushed. I really should retest after trivial refactorings...

**Comment by ghalliday** [comment]
@richardkchapman looks correct, and a worthwhile fix.

**Pull request merged**